### PR TITLE
Initial merge of tensorflow-training-docker

### PR DESCRIPTION
**Pull request by snormore**
Train a model in a Docker image using code and datasets fetched from storage URLs, pushing model output and logs to their own storage URLs.

[Build Status](https://travis-ci.org/triagemd/tensorflow-training-docker)

# Examples

## Keras trainer

```bash
git clone https://github.com/triagemd/keras-trainer.git
cd keras-trainer
source script/env
script/up
docker run triage/tensorflow-training:latest
  --volume "$(pwd):/trainer"
  --env TRAINING_PYTHON_PACKAGE="/trainer/dist/keras-trainer-0.0.1.tar.gz" \
  --env TRAINING_COMMAND="python -m keras_trainer.run mobilenet_v1 cat,dog /training/dataset/train /training/dataset/val /training/output/model /training/output/logs" \
  --env TRAINING_TRAIN_DATASET="/trainer/tests/files/catdog/train.zip" \
  --env TRAINING_VAL_DATASET="/trainer/tests/files/catdog/val.zip" \
  --env TRAINING_OUTPUT_MODEL="/trainer/output/model" \
  --env TRAINING_OUTPUT_LOGS="/trainer/output/logs"
```

Note that this example uses local storage, but URLs with `gs://` (auth via `GCLOUD_ACCOUNT` env) and `http(s)://` schemes are also supported.

### CODE REVIEW

File: examples/keras-trainer/keras_trainer/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+from .trainer import Trainer
```
Comment (snormore) [inline examples/keras-trainer/keras_trainer/__init__.py]: This example trainer is moving to https://github.com/triagemd/keras-trainer

File: script/distribute-docker-gpu [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+set -e
+VERSION=$(grep "version=" setup.py | tr -d "version='" | tr -d ",'" | tr -d "[:space:]")
+set -x
+docker build -t triage/tensorflow-training:$VERSION -f docker/gpu.Dockerfile .
+docker build -t triage/tensorflow-training:latest -f docker/gpu.Dockerfile .
```
Comment (stephensolis) [inline script/distribute-docker-gpu]: `tensorflow-training-gpu`

File: tensorflow_training/archive.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            self._extract_targz(self.file.name, output_dir)
+        elif self.file.name.lower().endswith('.zip'):
+            self._extract_zip(self.file.name, output_dir)
+        else:
+            raise ValueError('unknown archive format in %s, unable to extract' % (self.file.name, ))
+
+    def _extract_targz(self, archive_file, output_dir):
+        tar = tarfile.open(archive_file, 'r:gz')
+        with TemporaryDirectory() as temp_dir:
+            for member in tar.getmembers():
```
Comment (stephensolis) [inline tensorflow_training/archive.py]: You can replace all of this with `tar.extractall`

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,17 @@
+FROM triage/python3.6-tensorflow-optimized:latest
+
+ADD setup.py /tmp/app/
+ADD requirements.txt /tmp/app/
+RUN pip install --upgrade pip && \
+    cd /tmp/app && \
+    pip install -r requirements.txt
+
+RUN pip install Keras Pillow h5py
```
Comment (stephensolis) [inline Dockerfile]: Is this actually needed? Can't seem to find anywhere it gets used

Comment (stephensolis) [inline Dockerfile]: Is this actually needed? Can't seem to find anywhere it gets used

**Pull request merged**